---
title: "EDA for NYCDSA Shiny Project"
author: "Ryan Willett"
output: html_notebook
---

Stated objectives for this project submitted during the declaration are: 

"Analysis of RNA expression data on 6000 genes from 92 sets of yeast RNAseq next generation sequencing samples
• Clustering analysis to ascertain similarities in gene expression across experimental conditions or mutant backgrounds
• Display the significant differences in gene expression between experimental conditions or mutant backgrounds
• Identify features associated with enriched gene sets (e.g. subcellular localization, molecular function, etc)"

```{r, echo = F}
library(tidyverse)
library(ggplot2)
library(ggrepel)
library(dendextend)
library(ape)
```


Read in expression data from the data file. It is currently arranged to that the yeast strains (observations) are in the column and the gene names (variables) are in the row. I applied a transpose to switch the columns and rows.
```{r}
data <- read.csv(file="./data/SC_expression.csv", stringsAsFactors = F, row.names = 1, header = T)
conditions <- read.csv(file="./data/conditions_annotation.csv", stringsAsFactors = F, header = T)
data <- as.data.frame(t(data))
```


```{r}
# Inspecting that the data looks OK
data[1:20, 1:10]
```


Bringing in the condition data labels
```{r}
data <- data %>% rownames_to_column("ID") # Making ID column from the rownames so I can join with conditions table

data <- left_join(data, conditions, by="ID") # Binding the experimental conditions to the expression data

data$additional_information <- NULL # Don't need this column and it is mostly blank
```

Cleaning up the categories
```{r}
data <- data %>% 
  mutate(primary = gsub("wild type", "wildtype", primary)) %>% # Aligning the format of control name
  mutate(primary = gsub("^YB.*", "Biofuel.Production.Strain", primary)) %>%
  mutate(primary = gsub("^37.*", "37 deg", primary)) %>% # These were not grouped properly. Renamed to align
  mutate(primary = gsub("^Strain.*", "Phenol.Lysis", primary)) %>% 
  mutate(primary = gsub("^DDY.*", "Strain.W303", primary)) %>% 
  mutate(primary = gsub("^E.*", "Ethanol", primary)) %>% 
  mutate(primary = gsub("^G.*", "Glucose", primary)) %>% 
  mutate(primary = gsub("^hybrid.*", "hybrid cerevisiae x paradoxus", primary)) %>% select(-secondary) %>% 
  filter(!is.na(primary))


```


```{r}
# To explore the categories in the dataset
data %>% group_by(primary) %>% tally() 

levels(factor(data$primary))
```

Calculating the baseline statistics
```{r}
#Descriptive statistics

data_means <- data %>%
  select(-ID) %>% 
  group_by(primary) %>% 
  summarise_all(list(mean = mean)) %>% 
  column_to_rownames("primary") %>% 
  t() %>% as.data.frame()  # Transposing the df (converts to matrix) and reclassifying as a df

data_sds <- data %>% 
  select(-ID) %>% 
  group_by(primary) %>% 
  summarise_all(list(sd = sd)) %>% 
  column_to_rownames("primary") %>% 
  t() %>% as.data.frame()  # Transposing the df (converts to matrix) and reclassifying as a df

```

```{r}
write.csv(data, file = "./data/yeast_raw.csv", row.names = F)
write.csv(data_means, file = "./data/yeast_means.csv")
write.csv(data_sds, file = "./data/yeast_sds.csv")
```

```{r}
data <- read.csv("./data/yeast_raw.csv", stringsAsFactors = F, header = T)
data_means <- read.csv("./data/yeast_means.csv", stringsAsFactors = F, header = T, row.names=1)
```


Filtering with the inputs from the user
```{r}
input1 <- "wildtype"
input2 <- "Biofuel.Production.Strain"

#data_subcat <- data %>% filter(primary == input1 | primary == input2) # Filtering for the groups to be compared
```

Evaluation of the statistical difference between the selected groups. This will be between 2 groups, so it's a t-test
```{r}
p_eval <- function(x, y) {
  ttest_res <- t.test(x,y) # t-test against the 2 categories to determine significance
  p.val <- ttest_res$p.val # Extracting the p-value from the t.test output
  return(p.val)
}
```

```{r}

subcat_in1 <- data %>% filter(primary == input1)

subcat_in2 <- data %>% filter(primary == input2)

pstats <- c()

for (c in 2:(ncol(data)-1)){ 
  test1 <- subcat_in1[,c]
  test2 <- subcat_in2[,c]
  
  p <- p_eval(test1, test2)
  pstats[c] <- p # Build up p-values list into a vector
}
```

```{r}
data_means_sub <- data_means %>% select(input1, input2) %>% rownames_to_column("genes") %>% mutate(genes = gsub("_mean", "", genes))

data_means_sub <- cbind(data_means_sub, pstats=pstats[-1])

data_means_sub <- data_means_sub %>% 
  mutate(fold.change = data_means_sub[,3]/data_means_sub[,2],
         l2fc = (log2(data_means_sub[,3]) - log2(data_means_sub[,2])), 
         neg_pL10 = -1*log10(pstats)) %>% 
  mutate(l2fc = ifelse(l2fc == Inf, 10, ifelse(l2fc == -Inf, -10, l2fc))) %>% # Imputing in a floor and ceiling value to +/- Inf
  mutate(fold.change = ifelse(fold.change == Inf, 10, ifelse(fold.change == -Inf, -10, fold.change)))



```

Preparing the data for plotting
```{r}
data_means_sub <- data_means_sub %>% 
  mutate(col_flag = ifelse(((l2fc>=1 | l2fc<=-1) & neg_pL10 > 1.3),1,0)) 
```

```{r}
write.csv(data_means_sub, file = "./data/expression_data.csv", row.names = F)
```

```{r}
expr_data <- read.csv("./data/expression_data.csv", stringsAsFactors = F)
```

Preparing the data for plotting
```{r}
t <- theme_classic()
plot_colors <- scale_color_manual(values = c("black", "red", "green"))

n1 <- 0.05
n2 <- 2

data_means_sub %>% 
  mutate(col_flag = ifelse(((l2fc>=log2(n2) | l2fc<=-log2(n2)) & neg_pL10 > -1*log10(n1)),1,0)) %>% 
  ggplot(aes(x=l2fc, y=neg_pL10, col=as.factor(col_flag))) + t + 
  geom_point(size=0.2) + #+scale_y_continuous(trans="log10")
  plot_colors +
  #scale_y_continuous(trans = "log2") +
  labs(x = "log2 Fold Change in Gene Expression", y = "-log10(p-value)")
```

```{r}
sig_genes <- expr_data %>% 
  mutate(col_flag = ifelse(((l2fc>=1 | l2fc<=-1) & neg_pL10 > 1.3),1,0)) %>% 
  filter(col_flag == 1 & !is.na(fold.change)) %>% 
  select(-col_flag)

top_sig_genes <- expr_data %>% 
  mutate(col_flag = ifelse(((l2fc>=1 | l2fc<=-1) & neg_pL10 > 1.3),1,0)) %>% 
  filter(col_flag == 1 & !is.na(fold.change)) %>% 
  select(-col_flag) %>% 
  arrange(desc(fold.change)) %>% 
  slice(-101:-(n()-101)) #Return the highest 100 and lowest 100 gene expression changes
```

Heatmap
```{r}
hm_d <- data %>% filter(primary == input1 | primary == input2) %>% select(-primary) %>% t() %>% as.data.frame()
colnames(hm_d) <- as.character(unlist(hm_d[1,]))
hm_d <- hm_d[-1,]
hm_d <- hm_d %>% rownames_to_column("genes")

sigs <- top_sig_genes[,1, drop =F]
sigs_join <- merge(sigs, hm_d, by="genes")
sigs_join <- sigs_join %>% column_to_rownames("genes") %>% as.matrix()
sigs_join <- apply(sigs_join, 2, as.numeric) # Converting character matrix to numeric matrix for heatmap calculation
heatmap(sigs_join, scale = "row", Rowv = NA)
```


Clustering analysis
```{r}
data_raw <- data %>% select(-primary) # Removing categorical data so they are not used in the clustering

d <- dist(as.matrix(data_raw)) #cancer_data is used so as not to bias the clustering with the known tumor classification in cancer_join
cat_n <- 4
hc <- hclust(d, method="ward.D")
hc <- as.dendrogram(hc)

#Plot the data
#cols_branches <- c("blue", "green", "red", "black", "orange")
dend <- color_branches(hc, k=cat_n, col = c("blue", "green", "red", "black", "orange"))
#labels(dend) <- NULL # Remove labels because they make the image too busy
plot(dend, main = "Clustering Based on 4 Classes")


# circlize_dendrogram(dend)
plot(dend, type = "triangle", ylab = "Height")

dend <- dend %>% set("labels", "")
rad_dend_col <- c("blue", "green", "red", "black", "orange")
clus <- cutree(hc, 4)
plot(as.phylo(hc), type="fan", tip.color = rad_dend_col[clus], label.offset = 1, main = "Clustering Based on 4 Classes")

ggplot(dend) +
  coord_polar(theta = "x") +
  scale_y_reverse(expand = c(0.2, 0))
```



